apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SimpleSubscriber.cs

```diff
@@ -649,94 +650,125 @@ private class SingleChannel
                         {
                             _registerTaskFn(_taskHelper.Run(() => pull.WriteCompleteAsync()));
                         }
+                        streamingPullCts.Dispose();
                     }
                 }
             }
 
             private async Task Pull(SubscriberClient.StreamingPullStream pull, CancellationToken streamingPullToken)
             {
-                var cts = CancellationTokenSource.CreateLinkedTokenSource(streamingPullToken, _softStopCts.Token);
-                // .MoveNext() will only accept CancellationToken.None (by design seemingly), so we have to use the cancellationToken
-                // specified when the StreamingPull was started.
-                while (!cts.IsCancellationRequested && await _taskHelper.ConfigureAwait(pull.ResponseStream.MoveNext(CancellationToken.None)))
+                using (var cts = CancellationTokenSource.CreateLinkedTokenSource(streamingPullToken, _softStopCts.Token))
                 {
-                    // If in soft-stop, the overall StreamingPull might not yet be cancelled, but don't process any more msgs.
-                    if (cts.IsCancellationRequested)
+                    async Task<bool> MoveNext()
                     {
-                        break;
+                        // Pause pulling more messages if too many msgs are locally queued for sending.
+                        // The size of the extend queue is a reasonable proxy for push loading.
+                        while (_qLock.Locked(() => _extendQueue.Count) >= _maxAckExtendCount)
+                        {
+                            // A 100ms pause is fairly arbitrary, but should never cause problems.
+                            // Using an event would be better, but this is simpler and easier to assure correctness.
+                            await _taskHelper.ConfigureAwait(_scheduler.Delay(TimeSpan.FromMilliseconds(100), cts.Token));
+                        }
+                        return await pull.ResponseStream.MoveNext(cts.Token);
                     }
-                    var msgs = pull.ResponseStream.Current.ReceivedMessages;
-                    var ackIds = new HashSet<string>(msgs.Select(x => x.AckId));
-                    var allMsgsHandledCts = CancellationTokenSource.CreateLinkedTokenSource(_hardStopCts.Token);
-                    _registerTaskFn(_taskHelper.Run(() => AutoExtend(ackIds, allMsgsHandledCts.Token)));
-                    foreach (var msg in msgs)
+                    while (!cts.IsCancellationRequested && await _taskHelper.ConfigureAwaitHideCancellation(MoveNext, false))
                     {
                         if (cts.IsCancellationRequested)
                         {
-                            break;
+                            return;
```

Comment:
This is deliberately not throwing, because it's not the user's token, yes? (A comment might be worthwhile.)

Reply:
Correct. Comment added.